### ISSUE TITLE
List of directories in the "Search in Files" dialog - K-IDE 10.1.1

### ISSUE THREAD

**Issue by AvalonNZ** [Komodo/KomodoEdit#2012]
### Short Summary

I have half a dozen projects. It seems that every time I change projects (which is fequently) and I try to search for something, I have to go into the "Select directories in which to search for files" dialog box, remove the directory for the last project, and find the directory for the current project before I can start searching. This is really frustrating, especially since there's a perfectly good "recently used directories" dropdown list that only seems to remember one or two of the directories I've visited and seldom contains the one I need. The online documentation doesn't really talk much about this feature so I don't know if there's a setting somewhere that controls the number of files it will remember.

It would be easier to use if:
1. the recently used directories dropdown remembered more directories, or if this was easily configurable.
2. the directories listed in the "Select directories in which to search for files" dialog were automatically stored with the project settings, and were automatically restored when I change projects.
3. When presented with the "Select directories in which to search for files" dialog, it would be nice to be presented with a list of recently used directory collections to make using it faster.

Implementing ideas 1 and 2 would make 3 unnecessary. 

### Expected results

When I change to Project X, the directories I need to search within for Project X should be automatically restored for the "Select directories in which to search for files" dialog box.  

The recently used directories dropdown at the bottom of the screen should contain the last 8 or 9 directories I've searched in so it's easy to change the scope of the search (and this list of directories should contain multiple directories per line if the "Select directories in which to search for files" dialog box contained multiple directories).

### Actual results

When I change to Project X, the directories in the "Select directories in which to search for files" dialog are those I manually set up the last time I searched in files, so I have to manually change them before I can start searching. 

Furthermore, the recently used directories dropdown at the bottom of the screen only has a couple of directories in it and they don't usually contain the directories I need. 

### Platform Information

*Komodo IDE* 
*10.1.1* 
*Windows 10 Pro  - version 1511* 
64-bit, 8GB Ram

**Comment by Naatan** [comment]
Hi @AvalonNZ, could you please summarize your request in a few lines? The extra description is useful for context, but currently it makes it hard to boil down to the core issue.

**Comment by AvalonNZ** [comment]
Hi Naatan,

I guess the following covers it:

When I change to Project X, the directories I need to search within for Project X should be automatically restored for the "Select directories in which to search for files" dialog box (currently I get the directories used last time I searched, which was often for another project).

The recently used directories dropdown at the bottom of the screen (which currently shows only the last 2 or 3 directories I've searched in) should contain the last 8 or 9 directories (and this list of directories should contain multiple directories per line if the "Select directories in which to search for files" dialog box contained multiple directories).

Does that help?

Cheers,
Simon

**Comment by Naatan** [comment]
Thanks for explaining @AvalonNZ.

Unfortunately the way this is currently set up to use Mozilla API's it makes it difficult to change the behavior without refactoring the whole dropdown. Personally I'm not convinced this is worth doing as I've never really ran into this myself and have not seen from anyone else with this problem. I'll leave this open in the backlog for future consideration but don't see it worth targeting for a milestone at this time.

**Comment by AvalonNZ** [comment]
Hi Nathan,

I can appreciate the problem of rewriting the dropdown if the Mozilla APIs
make it difficult. It's not a show-stopper issue, and I probably wouldn't
invest the time yet either. I'm going to try using the "search in project"
feature to get around the problem. I hadn't realised that this feature
existed.

I'm curious as to whether you tried moving between half a dozen projects
with different directories, and tried to "search in files" to experience
the problem? At the very least I'd be keen to hear if you can replicate the
issue.

If I'm wrong, then maybe there's a setting (or an OS issue) that I'm not
aware of and maybe I can rectify the problem at my end. If I'm right (and
I've had this issue since I started using Komodo many years ago) then this
issue is likely to be one of those minor, workflow-interupting annoyances
that comes out when programmers talk to other programmers about their
favourite IDE. ie: "I really lke Komodo. The only annoyng thing about it
is...". Just a thought.

Cheers,
Simon

**Comment by Naatan** [comment]
I tend to search in files in 2 ways:

 * Search in Projects
 * Search from the breadcrumbs (Ctrl+Click a breadcrumb)

I rarely use the actual dropdown menu's from the search dialog as it is far more work than just using the shortcuts. So I've never ran into this issue myself no, though I can certainly see how someone could run into it.

**Comment by AvalonNZ** [comment]
Ah. Fair enough. I'll try the breadcrumb thing too. I didn't know that
search facility existed either.

tldr: jump to > below :-)

I write software (currently) to manage a pharmaceutical terminology. It
requires sophisticated searching techniques to allow users to locate
specific subsets of records via a wide variety of keys (similar to your
contexts) and partial matches, many of which return hundreds or even
thousands of results so I need to provide all sorts of filters.

I decided to try to emulate google search (which my users - all over 50
years of age - were comfortable using) and set up something like your
commando to make things easy. It automatically treats numeric search
criteria as one of three potential keys, and searches them in order of
likelihood based on the length of the criteria before defaulting to the
basic text search option. An exact hit on a numeric key delivers them
straight to the record in questoin without going through a search results
screen.

When the criteria is text, the search queries ignore the numeric indexes
for obvious reasons. My users seem to get the idea that they can type
anything into the search box and they'll get a result.

Some time back I discovered that they often wasted a lot of time selecting
one of the filters (typlically table names) from a drop-down box beside the
search box, and then clicking the search box to enter the criteria, and
then clicking the "go" button to activate the search.

I added a feature that allows them to simply type the table name (most of
which are acronyms that are four characters or less) as the first word in
the search criteria in order to specify the filter. They've always been
able to simply press enter rather than click the "go" button to activate
the search.

Despite the obvious utility and speed of the new approach, NOBODY USES IT.
They'd rather plod through the old method, as tedious as it is.

I've also added filtering methods to allow them to exclude specific terms
from the results using minus signs (which was inspired by the search
results filter mechanism you built into Komodo, but my filter is applied at
the front end rather than after the search because it's more efficient when
using a database). I suspect they only remember to use this for very
specific operations.

> My point is that it's hard to educate users to use new methods, and when
those mechanisms effectively hide their controls behind intelligent but
very simple interfaces (like commando) the job is even harder. This is
esepecially true when the old methods are still available (as they must
be).

It might be usefult to send out a series of "Did you know..." emails via
mailchimp (or whatever you use) to keep reminding people that there are
lots of new ways to do cool things faster. Maybe some videos? The Komodo
help system covers a lot of ground and it can be a little superfical in
places (including the search topic).

I love this product. It's the best IDE I've ever used and I really want it
to succeed and contine so I can keep using it for another 20 years. Keep up
the good work.

Cheers,
Simon

**Comment by Naatan** [comment]
Thanks Simon :) We do on occasion have screencasts and webinars where we walk through features. Additionally the new tutorial tool introduced in Komodo 10.1 will go a long way in explaining certain features to users.

At the end of the day users will use what is most obvious to them though, and will only start to discover new things when the need arises. Improving our general UX is what will facilitate this in the end. For example if you click a breadcrumb it will tell you that you can "Ctrl+Click" them in the dropdown menu.
